File: chapter-04/03_divisor_more_efficiently.c
#include <stdio.h> 
#include <math.h>
#include <string.h>

void array_sorting(int array[], int len) {
    int i, j, temp;
    for (i = 0; i < len; i++) {
        for (j = i+1; j < len; j++) {
            if (array[i] > array[j]) {
                temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }
        printf(" %d", array[i]);
    }
}

int main() {
    int num, i, j, k;

    while(1) {
        printf("Please enter a positive integer. Enter 0 to quit\n: ");
        scanf("%d", &num);
        if (num == 0) {
            break;
        }
        if (num < 0) {
            printf("You must enter a positive integer. Please try again.\n");
            continue;
        }
        int small_divisor[num];
        printf("Factors of %d : 1", num);
        for (i = 2, j = 0; i <= (sqrt(num))+1; i++) {
            if (num % i == 0) {
                small_divisor[j] = i;
                j++;
            }
        }
        for (k = j, i = 0; i < j; i++) {
            if (num / small_divisor[i] != small_divisor[i]) {
                small_divisor[k] = num / small_divisor[i];
                k++;
            }
        }
        array_sorting(small_divisor, k);
        // printf("\nk = %d\n", k);
        // for (i = 0; i < k; i++) {
        //     printf(" %d", small_divisor[i]);
        // }
        
        printf(" %d\n\n", num);
    }

    return 0;
}

File: gonit-korbo-joi_tamim-shahriar-subeen/chapter-04/03_divisor_more_efficiently.py
import math

while True:
    n = input("Please enter a positive integer. Enter 0 to quit: ")
    n = int(n)
    if n == 0:
        break
    if n < 0:
        print("You must enter a positive integer. Please try again.")
        continue

    factors = [1, n]

    sqrt_n = math.sqrt(n)
    sqrt_n = int(sqrt_n)
    
    for i in range(2, sqrt_n+1):
        if n % i == 0:
            factors.append(i)
            if (i != n//i):
                factors.append(n//i)
    
    factors = sorted(factors)
    print("Factors of", n, ":", factors)
    print("\n")